File: app/src/main/java/de/christinecoenen/code/zapp/app/mediathek/ui/detail/SelectQualityDialog.java
package de.christinecoenen.code.zapp.app.mediathek.ui.detail;

import android.app.Dialog;
import android.content.Context;
import android.content.DialogInterface;
import android.os.Bundle;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.appcompat.app.AlertDialog;
import androidx.appcompat.app.AppCompatDialogFragment;

import java.util.List;

import de.christinecoenen.code.zapp.R;
import de.christinecoenen.code.zapp.app.mediathek.model.MediathekShow;
import de.christinecoenen.code.zapp.app.mediathek.model.Quality;


public class SelectQualityDialog extends AppCompatDialogFragment {

	private static final String ARGUMENT_MEDIATHEK_SHOW = "ARGUMENT_MEDIATHEK_SHOW";
	private static final String ARGUMENT_MODE = "ARGUMENT_MODE";

	static SelectQualityDialog newInstance(MediathekShow mediathekShow, Mode mode) {
		SelectQualityDialog fragment = new SelectQualityDialog();

		Bundle args = new Bundle();
		args.putSerializable(ARGUMENT_MEDIATHEK_SHOW, mediathekShow);
		args.putSerializable(ARGUMENT_MODE, mode);
		fragment.setArguments(args);

		return fragment;
	}

	private List<Quality> qualities;
	private String[] qualityLabels;
	private Mode mode;
	private Listener listener;

	@Override
	public void onCreate(@Nullable Bundle savedInstanceState) {
		super.onCreate(savedInstanceState);

		assert getArguments() != null;

		mode = (Mode) getArguments().getSerializable(ARGUMENT_MODE);

		MediathekShow show = (MediathekShow) getArguments().getSerializable(ARGUMENT_MEDIATHEK_SHOW);
		assert show != null;

		switch (mode) {
			case DOWNLOAD:
				qualities = show.getSupportedDownloadQualities();
				break;
			case SHARE:
				qualities = show.getSupportedStreamingQualities();
				break;
		}

		qualityLabels = new String[qualities.size()];

		for (int i = 0; i < qualities.size(); i++) {
			qualityLabels[i] = getString(qualities.get(i).getLabelResId());
		}
	}

	@Override
	public void onAttach(@NonNull Context context) {
		super.onAttach(context);

		if (getTargetFragment() instanceof Listener) {
			listener = (Listener) getTargetFragment();
		} else {
			throw new IllegalArgumentException("Parent fragment must implement ConfirmFileDeletionDialog.Listener interface.");
		}
	}

	@NonNull
	@Override
	public Dialog onCreateDialog(@Nullable Bundle savedInstanceState) {
		return new AlertDialog.Builder(requireActivity())
			.setTitle(R.string.fragment_mediathek_qualities_title)
			.setItems(qualityLabels, this::onItemSelected)
			.setNegativeButton(android.R.string.cancel, null)
			.create();
	}

	private void onItemSelected(DialogInterface dialogInterface, int i) {
		Quality quality = qualities.get(i);

		switch (mode) {
			case DOWNLOAD:
				listener.onDownloadQualitySelected(quality);
				break;
			case SHARE:
				listener.onShareQualitySelected(quality);
				break;
		}
	}

	public enum Mode {
		DOWNLOAD,
		SHARE
	}

	interface Listener {
		void onDownloadQualitySelected(Quality quality);

		void onShareQualitySelected(Quality quality);
	}
}

File: app/src/main/java/de/christinecoenen/code/zapp/app/mediathek/ui/detail/dialogs/SelectQualityDialog.kt
package de.christinecoenen.code.zapp.app.mediathek.ui.detail.dialogs

import android.app.Dialog
import android.os.Bundle
import androidx.appcompat.app.AppCompatDialogFragment
import androidx.fragment.app.setFragmentResult
import com.google.android.material.dialog.MaterialAlertDialogBuilder
import de.christinecoenen.code.zapp.R
import de.christinecoenen.code.zapp.models.shows.MediathekShow
import de.christinecoenen.code.zapp.models.shows.Quality

class SelectQualityDialog : AppCompatDialogFragment() {

	companion object {

		const val REQUEST_KEY_SELECT_QUALITY = "REQUEST_KEY_SELECT_QUALITY"
		private const val REQUEST_KEY_SELECT_QUALITY_KEY_QUALITY =
			"REQUEST_KEY_SELECT_QUALITY_KEY_QUALITY"
		private const val ARGUMENT_MEDIATHEK_SHOW = "ARGUMENT_MEDIATHEK_SHOW"
		private const val ARGUMENT_MODE = "ARGUMENT_MODE"

		@JvmStatic
		fun newInstance(mediathekShow: MediathekShow, mode: Mode): SelectQualityDialog {
			return SelectQualityDialog().apply {
				arguments = Bundle().apply {
					putSerializable(ARGUMENT_MEDIATHEK_SHOW, mediathekShow)
					putSerializable(ARGUMENT_MODE, mode)
				}
			}
		}

		@JvmStatic
		fun getSelectedQuality(bundle: Bundle): Quality {
			return bundle.getSerializable(REQUEST_KEY_SELECT_QUALITY_KEY_QUALITY) as Quality
		}
	}

	private lateinit var mode: Mode
	private lateinit var qualities: List<Quality>
	private lateinit var qualityLabels: List<String>

	override fun onCreate(savedInstanceState: Bundle?) {
		super.onCreate(savedInstanceState)

		mode = requireArguments().getSerializable(ARGUMENT_MODE) as Mode

		val show = requireArguments().getSerializable(ARGUMENT_MEDIATHEK_SHOW) as MediathekShow

		qualities = when (mode) {
			Mode.DOWNLOAD -> show.supportedDownloadQualities
			Mode.SHARE -> show.supportedStreamingQualities
		}

		qualityLabels = qualities.map { getString(it.labelResId) }
	}

	override fun onCreateDialog(savedInstanceState: Bundle?): Dialog {
		return MaterialAlertDialogBuilder(requireActivity())
			.setTitle(R.string.fragment_mediathek_qualities_title)
			.setItems(qualityLabels.toTypedArray()) { _, i ->
				onItemSelected(qualities[i])
			}
			.setNegativeButton(android.R.string.cancel, null)
			.create()
	}

	private fun onItemSelected(quality: Quality) {
		val bundle = Bundle().apply {
			putSerializable(REQUEST_KEY_SELECT_QUALITY_KEY_QUALITY, quality)
		}

		setFragmentResult(REQUEST_KEY_SELECT_QUALITY, bundle)
	}

	enum class Mode {
		DOWNLOAD, SHARE
	}
}
